src/common/device_helpers.cuh

```diff
@@ -866,16 +869,30 @@ class AllReducer {
    * group.
    *
    * \param device_ordinals The device ordinals.
+   * \param opg whether to use nccl one-process-per-gpu type initialization
    */
 
-  void Init(const std::vector<int> &device_ordinals) {
+  void Init(const std::vector<int> &device_ordinals, bool opg=false) {
 #ifdef XGBOOST_USE_NCCL
+    this->use_nccl_opg = opg;
+    
     /** \brief this >monitor . init. */
+    
     this->device_ordinals = device_ordinals;
     comms.resize(device_ordinals.size());
-    dh::safe_nccl(ncclCommInitAll(comms.data(),
-                                  static_cast<int>(device_ordinals.size()),
-                                  device_ordinals.data()));
+    if (use_nccl_opg) {
```

Comment:
Instead of having an if statement here, I would like you to initialise using ncclCommInitRank but generalise such that it also works with multiple GPUs per process then remove the second part of this if statement.

Reply:
`ncclCommInitRank(ncclComm_t* newcomm, int nranks, ncclUniqueId commId, int myrank)` initializes the rank of each GPU in a distinct process.
`ncclCommInitAll(ncclComm_t* comms, int ndev, const int* devlist)` initializes the ranks of all GPUs in `devlist` on the system in a single process (provided that `ndev` is less than or equal to the length of `devlist` ... if less, then it will only initialize the first `ndev` GPUs).

To accomplish this, `ncclCommInitAll` uses `sched_getaffinity` and `sched_setaffinity` to control the CPU mask, limiting initialization to a single PID (the calling PID `0`).

It may be possible to wrap the call to `ncclCommInitRank` with `sched_setaffinity` in a similar fashion to force all initialized GPU ranks to be under a single PID; though, I'm not sure I like this design as it requires more manual (un)setting of the PID controlling the GPU ranks.

It is also possible that I've grossly misinterpreted your ask. If so, please correct.